File: apps/sdkwork-rtc-flutter-mobile/packages/sdkwork_rtc_flutter_mobile_admin_core/lib/src/services/provider_plugin_service.dart
import '../admin_sdk_mapper.dart';
import '../backend_rtc_client.dart';
import '../models/paginated_list_result.dart';
import '../models/provider_plugin.dart';

class ProviderPluginService {
  final SdkworkBackendClient _client;

  ProviderPluginService(this._client);

  Future<PaginatedListResult<ProviderPluginDescriptor>> list({
    int? page,
    int? limit,
    String? cursor,
    String? search,
    String? sort,
  }) async {
    final response = await _client.rtcProviderPlugins.list();
    return PaginatedListResult(
      items: backendResponseItems(response)
          .map(ProviderPluginDescriptor.fromJson)
          .toList(),
    );
  }

  Future<ProviderPluginDescriptor> get(String provider) async {
    final response = await _client.rtcProviderPlugins.retrieve(provider);
    return ProviderPluginDescriptor.fromJson(
      backendResponseEntity(response, 'Provider plugin $provider was not found'),
    );
  }
}

File: apps/sdkwork-rtc-pc/packages/sdkwork-rtc-pc-admin-core/src/services/providerPluginService.ts
import type { AuthTokenManager } from "@sdkwork/sdk-common";

import type { ProviderPluginDescriptor } from "../types/providerSchema";
import { resolveBackendRtcClient, type RtcBackendClientOptions, type RtcBackendClientSource } from "./backendClient";

interface ListResponse {
  items: ProviderPluginDescriptor[];
  nextCursor?: string | null;
}

export class ProviderPluginService {
  private readonly client;

  constructor(
    baseUrlOrClient: RtcBackendClientSource,
    tokenManagerOrOptions?: AuthTokenManager | RtcBackendClientOptions,
  ) {
    this.client = resolveBackendRtcClient(baseUrlOrClient, tokenManagerOrOptions);
  }

  async list(params?: {
    page?: number;
    limit?: number;
    cursor?: string;
    search?: string;
    sort?: string;
  }): Promise<ListResponse> {
    const response = await this.client.rtcProviderPlugins.rtc.providerPlugins.list({
      page: params?.page,
      pageSize: params?.limit,
      cursor: params?.cursor,
      q: params?.search,
      sort: params?.sort,
    });
    const data = response.data as
      | { items?: ProviderPluginDescriptor[]; nextCursor?: string | null }
      | undefined;
    return {
      items: data?.items ?? [],
      nextCursor: data?.nextCursor ?? null,
    };
  }

  async get(provider: string): Promise<ProviderPluginDescriptor> {
    const response = await this.client.rtcProviderPlugins.rtc.providerPlugins.retrieve(provider);
    const data = response.data as unknown as ProviderPluginDescriptor | undefined;
    if (!data) {
      throw new Error(`RTC provider plugin not found: ${provider}`);
    }
    return data;
  }
}
